src/__tests__/test-utils-selectors.test.tsx

```diff
@@ -0,0 +1,141 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import fs from 'fs';
+import Path from 'path';
+import glob from 'glob';
+import { transformSync, types, PluginObj, NodePath } from '@babel/core';
+import { flatten, zip } from 'lodash';
+
+// The test extracts generated selectors from the compiled output and matches those against the snapshot.
+test('test-utils selectors', () => {
+  // Find referenced selector files and properties.
+  const selectorsFilePathToUsedProperties = new Map<string, Set<string>>();
+  for (const file of glob.sync('lib/components/test-utils/selectors/**/*.js')) {
+    extractSelectorProperties(file, (filePath, propertyKey) => {
+      const properties = selectorsFilePathToUsedProperties.get(filePath) ?? new Set();
+      properties.add(propertyKey);
+      selectorsFilePathToUsedProperties.set(filePath, properties);
+    });
+  }
+
+  // Find referenced selector values.
+  const componentToSelectors: Record<string, string[]> = {};
+  for (const [filePath, properties] of selectorsFilePathToUsedProperties) {
+    extractComponentSelectors(filePath, [...properties], selector => {
+      const componentName = getComponentNameFromFilePath(filePath);
+      componentToSelectors[componentName] = [...(componentToSelectors[componentName] ?? []), selector].sort();
+    });
+  }
+  expect(componentToSelectors).toMatchSnapshot();
+});
+
+function extractSelectorProperties(file: string, onExtract: (filePath: string, propertyKey: string) => void) {
+  function extractor(): PluginObj {
+    const selectorVars = new Map<string, string>();
+    return {
+      visitor: {
+        // Find require statements that import selectors.
+        CallExpression(path: NodePath<types.CallExpression>) {
+          if (path.node.callee.type === 'Identifier' && path.node.callee.name === 'require') {
+            const argument = path.node.arguments[0];
+            if (argument.type === 'StringLiteral' && argument.value.endsWith('selectors.js')) {
+              if (path.parent.type === 'VariableDeclarator' && path.parent.id.type === 'Identifier') {
+                selectorVars.set(path.parent.id.name, Path.resolve(file, '..', argument.value));
+              }
+            }
+          }
+        },
+        // Find selector references and extract used property names.
+        MemberExpression(path: NodePath<types.MemberExpression>) {
+          function parseNested(expression: types.MemberExpression, propertyName: string) {
+            if (expression.object.type === 'Identifier' && selectorVars.has(expression.object.name)) {
+              const filePath = selectorVars.get(expression.object.name)!;
```

Comment:
This fails if it ever hits this code


```js
import styles from '...../selectors.js'

findSomething() {
   const styles = getSomethingElse();
   return this.find(styles[....]); // this will be incorrectly classified as coming from the import
}
```

Babel has functionality to track variables and where they are coming from: https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md#toc-scopes

Use scopes to track this properly

Reply:
This actually fails in the described scenario when running the script against uncompiled sources (in the compiled ones the var names are generated). However, in the worst case scenario the output would contain a few extra selectors which is fine. The case is extremely unlikely though: not only the variable name has to be duplicated, the accessed property must exist. I propose to not introduce scopes tracking unless there is a simple way of adding it.

Reply:
This is what you needed to do


```js
export default function (babel) {
  const { types: t } = babel;

  return {
    name: "ast-transform", // not required
    visitor: {
      MemberExpression(path) {
        const binding = path.scope.getBinding(path.node.object.name);
        if (binding.kind !== 'module' || !binding.path.parent.source.value.endsWith('selectors.js')) {
          return;
        }
        // this is an expected import
      }
    }
  };
}
```

Live demo: https://astexplorer.net/#/gist/9fcb7096d184ca40423225db5f41baeb/1f4d1737b0dfd6fb7fa025ec9da416f3dd1a1f93

Do you still think it is difficult?

Reply:
Many thanks for the reference, I was not familiar with getBinding API. I will add the code